docs(lockup-contract): refactored README for better clarity

telezhnaya opened the pull request:
What have been done:
- Added info about the difference between lockup and vesting, and how do we combine these logic
- Added info about lockup fields
- Added info about foundation account (it could be trird-party)
- Removed `technical details` section. All the info from it is outdated (first point) or written in corresponding sections.

Resolves #140, resolves #141

bowenwang1996 commented on lockup/README.md:
```diff
@@ -1,43 +1,64 @@
-This contract acts as an escrow that locks and holds an owner's tokens for a lockup period. A lockup period starts
-from the moment transfers are enabled by voting and lasts for the specified duration.
+This contract acts as an escrow that locks and holds an owner's tokens for a lockup period. 
+The contract contains of lockup and vesting processes that go simultaneously.
```
"contains of" is not correct English grammar. You might have meant "consists of"

bowenwang1996 commented on lockup/README.md:
```diff
@@ -1,43 +1,64 @@
+The contract contains of lockup and vesting processes that go simultaneously.
+
+A lockup period starts from the specified timestamp (or, if not specified, from moment when transfers are enabled by voting) 
+and lasts for the specified duration.
```
As suggested by @matklad, we start to use one sentence per line https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line

matklad replied:
@bowenwang1996 do we have some style.md somewhere where we could dump this info? A-la https://github.com/rust-analyzer/rust-analyzer/blob/d4a51241ceedae25f15bb8d202493554ae0af810/docs/dev/style.md#documentation ?

bowenwang1996 commented on lockup/README.md:
```diff
@@ -1,43 +1,64 @@
-The contract can contain a vesting schedule and serve as a vesting agreement between the NEAR Foundation (Foundation) and an employee (owner of contract).
+The contract can contain a vesting schedule and serve as a vesting agreement between the Foundation and an employee (owner of contract).
+
+Foundation is either the NEAR Foundation, or it could be third-party account. 
```
if so, is `foundation` still the best name?

frol replied:
@telezhnaya `foundation_account_id` is an unfortunate internal naming, but we still can use better wording on the documentation level and reference the internal naming explicitly pointing out where it comes from.

frol commented on lockup/README.md:
```diff
@@ -1,43 +1,64 @@
+
+A lockup period starts from the specified timestamp (or, if not specified, from moment when transfers are enabled by voting) 
+and lasts for the specified duration.
 It's also possible to lock tokens until the absolute timestamp. In this case the tokens will be available after the
```
> It's also possible to lock tokens until the absolute timestamp.

I don't believe that is true.

telezhnaya replied:
I think it's true, you could set `lockup_timestamp` to desired absolute timestamp and set `release_duration` to 0.
But it looks like this sentence adds nothing to the meaning, it could be deleted not to confuse anyone

frol commented on lockup/README.md:
```diff
@@ -1,43 +1,64 @@
-When the transfer poll is resolved, it returns the timestamp when it was resolved and it's used as the beginning of the
-lockup period.
+Vesting is an additional mechanism. It also locks the tokens, and it allows to configure 2 more options:
+1. Ability of terminating the contract from the Foundation.
```
```suggestion
1. Ability to terminate tokens vesting and refund non-vested tokens back.
```

frol commented on lockup/README.md:
```diff
@@ -1,43 +1,64 @@
-lockup period.
+Vesting is an additional mechanism. It also locks the tokens, and it allows to configure 2 more options:
+1. Ability of terminating the contract from the Foundation.
+2. Cliff: the period when all tokens remain locked. After the cliff, there is a linear unlock till the end of the contract.
```
Maybe something like this:

```suggestion
2. Cliff vesting period. All tokens remain non-vested before the cliff checkpoint is reached; on the cliff checkpoint, the accumulated tokens are vested in bulk; after the cliff, tokens are linearly unlocked till the vesting reaches its end timestamp.
```

or, just:

```suggestion
2. Cliff vesting period.
```

(the details are provided below)

frol commented on lockup/README.md:
```diff
@@ -1,43 +1,64 @@
+
+The contract could also have only one of these mechanisms. When initializing, it's possible to pass empty vesting information. 
+It's also possible not to provide `release_duration`, it means that we are using vesting schedule.
+If both mechanisms are not initialized, the tokens will be locked forever.
```
```suggestion
If neither of the mechanisms is initialized, the tokens will be locked forever.
```

telezhnaya commented:
I also created additional doc with cute pictures: https://github.com/telezhnaya/docs/blob/master/docs/tokens/lockup_vesting_details.md

frol reviewed (commented):
Overall, looks accurate to me. Once we have the visualizations, it will be perfect

frol commented:
BTW, I suggest using https://grammarly.com/ to check the grammar :wink:

evgenykuzyakov commented on lockup/README.md:
```diff
@@ -1,107 +1,110 @@
+### Combining lockup and vesting
 
-The release of tokens can also be scheduled to be linear from the moment tokens are unlocked.
-The tokens are unlocked at the following timestamp `max(transfers_enabled_timestamp + lockup_duration, lockup_timestamp)`.
```
Why this is removed? If both `lockup_duration` and `lockup_timestamp` is present, then we still should use the maximum of two. So with `cliff_timestamp`, we have:
```
max(max(transfers_enabled_timestamp + lockup_duration, lockup_timestamp), cliff_timestamp)
```

evgenykuzyakov commented on lockup/README.md:
```diff
@@ -1,107 +1,110 @@
+It is described by the following fields:
+- `lockup_timestamp` - The moment when the tokens become unlock (start of the contract).
+- `lockup_duration` - Alternative way to set the moment when the tokens become unlock.
+  The duration from the moment the transfers are enabled (`1602614338293769340`) to the start of the contract.
```
`TransferEnabled` timestamp can also be passed directly, so it might be different.

frol commented on lockup/README.md:
```diff
@@ -1,107 +1,110 @@
-from the moment transfers are enabled by voting and lasts for the specified duration.
-It's also possible to lock tokens until the absolute timestamp. In this case the tokens will be available after the
-lockup period or after the timestamp, whichever is later.
+This contract acts as an escrow that locks and holds an owner's tokens for a lockup period.
```
Let's add a link to https://docs.near.org/docs/tokens/lockup somewhere

nearmax commented on lockup/README.md:
```diff
@@ -1,107 +1,110 @@
 
-If there is still a termination balance deficit due to minimum required balance, the owner may decide to fund the deficit on this account to finish the termination process.
-This can be done through a regular transfer action from an account with liquid balance.
+If there is still a termination balance deficit due to the minimum required balance, the owner may decide to fund the deficit on this account to finish the termination process.
```
This sounds like an important mechanism that everyone should be aware of. Is it possible to pre-fund all lockup contracts upon creation with 35 extra NEAR to avoid having this situation upon termination?

telezhnaya replied:
https://github.com/near/core-contracts/issues/145

nearmax commented on lockup/README.md:
```diff
@@ -1,107 +1,110 @@
-  - If the NEAR Foundation account ID is provided during initialization, the NEAR Foundation can terminate vesting schedule.
-  - If the NEAR Foundation account ID is not provided, the vesting schedule can't be terminated.
-- Lock tokens for the lockup period with the release duration. Tokens are linearly released on transfers are enabled.
+Since the vesting schedule usually starts at the date of employment it allows to de-anonymize the owner of the lockup contract.
```
To clarify, this contract will be used not just for the employees.

telezhnaya commented on lockup/README.md:
```diff
@@ -1,107 +1,110 @@
+
+Lockup is a mechanism of linear unlocking of tokens that could not be terminated.
+It is described by the following fields:
+- `lockup_timestamp` - The moment when the tokens become unlock (start of the contract).
```
Sorry for the confusing phrasing.
Took your version except example in p.1. The idea of the timestamp (as I see it) is that we can forget about Mainnet launch, so I don't think this example is relevant right now.

telezhnaya commented on lockup/README.md:
```diff
@@ -1,107 +1,110 @@
-In addition to the lockup period that starts from the moment the transfers are enabled, vesting schedule also locks
-all tokens until `cliff_timestamp`.
-Once the `cliff_timestamp` passed, the tokens are vested on a pro rata basis from the `start_timestamp` to the `end_timestamp`.
+Once the `cliff_timestamp` passed, the tokens are vested on a pro-rata basis from the `start_timestamp` to the `end_timestamp`.
```
Are you sure that example above is not enough? (5 lines before)
Don't want to overload the doc.

We also have many examples [here](https://github.com/telezhnaya/docs/blob/master/docs/tokens/lockup_vesting_details.md), it will be merged into [our doc](https://docs.near.org/docs/tokens/lockup), I think we don't need additional example here.

telezhnaya commented on lockup/README.md:
```diff
@@ -1,107 +1,110 @@
-The tokens are unlocked at the following timestamp `max(transfers_enabled_timestamp + lockup_duration, lockup_timestamp)`.
+The contract could have both lockup and vesting schedules.
+
+The tokens are started to unlock at the following timestamp `max(max(transfers_enabled_timestamp + lockup_duration, lockup_timestamp), cliff_timestamp)`.
```
They work separately, but the resulting schedule is affected by both these processes.

telezhnaya commented on lockup/README.md:
```diff
@@ -1,107 +1,110 @@
+The current amount of locked tokens are calculated as the maximum between lockup and vesting logic.
+If at least one mechanism said the tokens are locked, then they are still locked.
+
+The contract could also have only one of these mechanisms. When initializing, it's possible to pass empty vesting information.
```
`vesting_schedule` is optional field https://github.com/telezhnaya/core-contracts/blob/master/lockup/src/lib.rs#L188
just pass `None`

> Does it mean that we only have lockup mechanics?

Yes

telezhnaya commented on lockup/README.md:
```diff
@@ -1,107 +1,110 @@
+
+The contract could also have only one of these mechanisms. When initializing, it's possible to pass empty vesting information.
+It's also possible not to provide `release_duration`, it means that we are using a vesting schedule.
+If neither of the mechanisms is initialized, the tokens will be locked forever.
```
Great catch, thank you! Rewriting this

telezhnaya commented on lockup/README.md:
```diff
@@ -1,107 +1,110 @@
+If the vesting is terminated before the cliff all tokens are refunded to the Contract Creator. Otherwise, the remaining unvested tokens are refunded.
 
-In the event of termination, the vesting stops and the remaining unvested tokens are locked until they are withdrawn by the Foundation.
+In the event of termination, the vesting stops, and the remaining unvested tokens are locked until they are withdrawn by the Contract Creator.
```
I can change it but I try to be not too wordy. I think that my wording gives the same amount of information, doesn't it?

telezhnaya commented on lockup/README.md:
```diff
@@ -1,107 +1,110 @@
 During termination, the owner can't issue any action towards the staking pool or issue transfers.
-If the amount of tokens on the contract account is less than the remaining unvested balance, the Foundation will try to unstake and withdraw everything from the staking pool.
-Once the tokens are withdrawn from the staking pool, the Foundation will proceed with withdrawing the unvested balance from the contract.
+If the amount of tokens on the contract account is less than the remaining unvested balance, the Contract Creator will try to unstake and withdraw everything from the staking pool.
```
It is possible if tokens are on the staking pool

frol commented on lockup/README.md:
```diff
@@ -1,107 +1,117 @@
-When the transfer poll is resolved, it returns the timestamp when it was resolved and it's used as the beginning of the
-lockup period.
+The owner can add a full access key to the account if all conditions are met:
+- Transfers are enabled;
```
This way it is much easier to comprehend :+1: 

I suggest providing more context whenever possible, so here we can provide the links to the initial voting plan and clarify that is already done:

```suggestion
- [Transfers are enabled](https://near.org/blog/near-mainnet-is-now-community-operated/) ([Phase II launch on MainNet enabled transfers](https://near.org/blog/near-mainnet-phase-2-unrestricted-decentralized/));
```

frol commented on lockup/README.md:
```diff
@@ -1,107 +1,117 @@
-lockup period or after the timestamp, whichever is later.
+This contract acts as an escrow that locks and holds an owner's tokens for a lockup period.
+The contract consists of lockup and vesting processes that go simultaneously.
+High-level overview could be found [in Near Docs](https://docs.near.org/docs/tokens/lockup).
```
```suggestion
A high-level overview could be found [in NEAR documentation](https://docs.near.org/docs/tokens/lockup).
```

frol commented on lockup/README.md:
```diff
@@ -1,107 +1,110 @@
 
-In case of vesting schedule, the contract supports the ability for the NEAR Foundation to terminate vesting at any point before it completes.
-If the vesting is terminated before the cliff all tokens are refunded to the Foundation. Otherwise the remaining unvested tokens are refunded.
+In the case of vesting schedule, the contract supports the ability for the Contract Creator to terminate vesting at any point before it completes.
```
I believe the confusion here is that "Contract Creator" is ambiguous since on mainnet it is lockup-factory which creates and deploys the contract. My suggestion would be to rename it to "foundation" (without "NEAR" and lowercase) implying any foundation which might be interested in founding lockup contracts and thus become the entity ensuring the trust to terminate vesting early. Our error messages in the contract return messages like:

> "Foundation account should be added for vesting schedule"

> "Foundation account can't be added without vesting schedule"

> "Can only be called by NEAR Foundation"

(the last one is unfortunately too specific)

telezhnaya commented:
Some changes were done. I cleaned up the history of commits, I also deleted changes from the code (I fixed some comments there, me and @frol decided not to touch the source code in this PR). 

The first commit is the original version + corrections after all reviews. In the second commit, I re-ordered things a little and marked `lockup_duration` and private vesting as deprecated.

frol reviewed (approved):
With this change and the change to the high-level overview in the NEAR docs (https://github.com/near/docs/pull/691), I believe, we reached the following:

* [x] Confirmed that the implementation itself is correct
* [x] The descriptions that were too low-level (reference-like) were extended with more context and references to the high-level ideas
* [x] The descriptions that were too generic (leading to uncertainty) were extended with specific examples and illustrations

@telezhnaya Thanks for the thorough review of all the information we had and driving this initiative forward!

nearmax commented on lockup/README.md:
```diff
@@ -1,107 +1,110 @@
-The owner can choose the staking pool where to delegate tokens.
-The staking pool contract and the account has to be approved and whitelisted by the foundation, to prevent lockup tokens from being lost, locked or stolen.
+The owner can choose the staking pool for delegating tokens.
+The staking pool contract and the account have to be approved and whitelisted by the NEAR Foundation, to prevent lockup tokens from being lost, locked, or stolen.
```
> It looks so. @evgenykuzyakov please confirm

Waiting for confirmation from @telezhnaya  or @evgenykuzyakov 

> I am not sure but it looks like it's not possible to pass the address of the whitelisting contract

It would be good if we could figure it out from the contract code. If whitelisting contract is hardcoded, then it would be good to have a link to the code where it is defined.

evgenykuzyakov replied:
> The staking pool contract and the account have to be approved and whitelisted by the NEAR Foundation, to prevent lockup tokens from being lost, locked, or stolen.

This is guaranteed by the whitelist contract account ID. It doesn't have to be chosen by NEAR Foundation, but if reused for other lockups, the guarantee stands.

This question is completely independent from the `foundation_account_id` meaning.
The `foundation_account_id` can be renamed to `terminator_account_id`, because the only function of this account is to be able to terminate vesting schedule.

nearmax commented on lockup/README.md:
```diff
@@ -1,107 +1,110 @@
 This staking pool must be an approved account, which is validated by a whitelisting contract.
-Once the staking pool holds tokens, the owner of the staking pool is able to use them to vote on the network governance issues, such as enabling transfers.
-So it's important for the owner to pick the staking pool that fits the best.
+Once the staking pool holds tokens, the owner of the staking pool can use them to vote on the network governance issues, such as enabling transfers.
```
Is it hardcoded in the contract? Could we please link the place from the source code in the documentation?

evgenykuzyakov replied:
I think this part is obsolete and should be deprecated. Since the voting for transfers has already passed there is no reason for the new lockup contracts to be initialized without explicitly enabled transfers with the fixed timestamp.

The account_id that was checking and counting votes is not hardcoded, but a part of the enum:
https://github.com/near/core-contracts/blob/d84f7a924b79009d3f7aa4e577c07b08a0489dc9/lockup/src/types.rs#L62

I think we shouldn't use this anymore and should explicitly mark it as deprecated in the code.

telezhnaya replied:
I created a separate issue about marking as deprecated/fixing comment in the contract: https://github.com/near/core-contracts/issues/146
Please comment there anything that you remember that should be fixed

nearmax reviewed (approved):
Thanks for the documentation fix, approving assuming the remaining comments are resolved.

frol commented:
I had to merge ASAP since it is good enough now, and I wanted to remove the scary banner and let partners review the contract without misleading messages. @telezhnaya please, submit a separate PR addressing the concerns from the comments

telezhnaya commented:
@frol thank you!
I continued [here](https://github.com/near/core-contracts/pull/147)
